SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/UNR-3761/SpatialTestHandover/SpatialTestHandover.cpp

```diff
@@ -0,0 +1,166 @@
+	MoveCubeStepDefinition.bIsNativeDefinition = true;
+	MoveCubeStepDefinition.NativeStartEvent.BindLambda([this]()
+		{
+			if (LocationIndex >= TestLocations.Num() || LocationIndex < 0)
+			{
+				return;
+			}
+
+			if (HandoverCube->HasAuthority())
+			{
+				HandoverCube->SetActorLocation(TestLocations[LocationIndex]);
+			}
+
+			++LocationIndex;
+			FinishStep();
+		});
+
+	// Step Definition to check if the correct server is authoritative over the HandoverCube.
+	FSpatialFunctionalTestStepDefinition CheckAuthorityStepDefinition;
+	CheckAuthorityStepDefinition.bIsNativeDefinition = true;
+	CheckAuthorityStepDefinition.TimeLimit = 10.0f;
+	CheckAuthorityStepDefinition.NativeTickEvent.BindLambda([this](float DeltaTime)
+		{
+			// If a Server Worker holds a lock over the HandoverCube, then it should have authority over it, if not, let the Load Balacning decide.
+			int ExpectedAuthoritativeServer = HandoverCube->LockingServerID ? HandoverCube->LockingServerID : LoadBalancingStrategy->WhoShouldHaveAuthority(*HandoverCube);
+			// Make sure the correct Server has authority over the HandoverCube.
+			if (GetLocalFlowController()->WorkerDefinition.Id == ExpectedAuthoritativeServer)
+			{
+				if (HandoverCube->HasAuthority())
+				{
+					FinishStep();
+				}
+			}
+			else
+			{
+				// Also ensure that all other Servers are not authoritative over the HandoverCube.
+				if (!HandoverCube->HasAuthority())
+				{
+					FinishStep();
+				}
+			}
+		});
+
+	// Server 1 spawns the HandoverCube under its authority area.
+	AddStep(TEXT("SpatialTestHandoverServer1SpawnCube"), FWorkerDefinition::Server(1), nullptr, [this]()
+		{
+			HandoverCube = GetWorld()->SpawnActor<AHandoverCube>(SpawnLocation, FRotator::ZeroRotator, FActorSpawnParameters());
+			RegisterAutoDestroyActor(HandoverCube);
+
+			FinishStep();
+		});
+
+	// All servers set a reference to the HandoverCube and reset the LocationIndex and AuthorityCheckIndex.
+	AddStep(TEXT("SpatialTestHandoverAllServersSetupStep"), FWorkerDefinition::AllServers, nullptr, nullptr, [this](float DeltaTime)
+		{
+			TArray<AActor*> HandoverCubes;
+			UGameplayStatics::GetAllActorsOfClass(GetWorld(), AHandoverCube::StaticClass(), HandoverCubes);
+
+			if (HandoverCubes.Num() == 1)
+			{
+				HandoverCube = Cast<AHandoverCube>(HandoverCubes[0]);
+
+				USpatialNetDriver* NetDriver = Cast<USpatialNetDriver>(GetWorld()->GetNetDriver());
+
+				AssertTrue(IsValid(NetDriver), TEXT("This test should be run with Spatial Networking"));
+
+				LoadBalancingStrategy = Cast<ULayeredLBStrategy>(NetDriver->LoadBalanceStrategy);
+				
+				if (IsValid(HandoverCube) && IsValid(LoadBalancingStrategy))
+				{
+					// Reset the LocationIndex to allow multiple executions of the test.
+					LocationIndex = 0;
+					FinishStep();
+				}
+			}
+		}, 5.0f);
+
+	// Check that Server 1 is authoritative over the HandoverCube.
+	AddStepFromDefinition(CheckAuthorityStepDefinition, FWorkerDefinition::AllServers);
```

Comment:
One concern I have here is that if the actor move command, doesn't actually move the cube to a new server authoritative area, then no handover will take place, but the test will all still pass.
Can we add some validation that handover is actually occurring? Perhaps tap into OnAuthGained/Lost?

Reply:
Or when we validate authority, we validate authority onto an expectant worker, that increments with each step

Reply:
Oh, yes this is a consequence of querying the load balancer that I did not take into account, will update this, thank you very much for the catch!